### ISSUE TITLE
New timed text page

### ISSUE THREAD

**Issue by FayCross** [thexerteproject/xerteonlinetoolkits#313]
New timed text page to allow more flexibility & options than the bullets page currently does (combines issues #40 & #211)

**Comment by JohnSmith-LT** [comment]
Hi @FayCross and @ronm123 

Does you think this commit is a useful addition to the bullets page, addresses some of the issues mentioned above and more importantly doesn't break any older LOs that use the bullets page (or at least don't change the functionality too far towards the unacceptable)?

It shouldn't modify the behaviour of any content that doesn't already include <li> and it can handle both ordered and unordered lists and separate lists with <p> blocks before, after and between.

![bullets](https://user-images.githubusercontent.com/3907362/69010778-017f1100-095b-11ea-9b76-5b8f18503028.gif)

**Comment by ronm123** [comment]
@JohnSmith-LT looks good from your animation. I'm probably not going to get a chance to test until Tuesday now.

**Comment by ronm123** [comment]
@JohnSmith-LT just testing this and I wonder if we should add a tooltip either to the Page Text or the Add Bullets to Text labels. e.g.
If you want a mix of normal paragraph text which doesn't have bullets and bulleted text that does have bullets then apply the bullets to relevant text using the wysiwyg editor. If you apply bullets in this way then the option below to add bullets to text will be ignored. However if you wish to automatically add bullets to all paragraphs do not add bullets via the wysiwyg editor and select the option to Add Bullets to Text.

Also looking at your gif and my own local example I'm wondering about the fact that bullets.css is adding padding-bottom:10px; to li.bullet which results in different spacing between Line 1 and Sub bullet 1 vs the spacing between Sub bullet 1 and Sub bullet 2. changing li.bullet to padding-top:10px; rather than padding-bottom:10px looks better to me - what do you think?

**Comment by JohnSmith-LT** [comment]
Thanks @ronm123 - whatever you think looks best with the padding is fine by me - alternatively (or additionally) it might need some css on the <ul> and <ol> tags to look right.

I'll add the tooltip later so if you need want to update the css at the same time then let me know what you've added that looks best.

**Comment by ronm123** [comment]
@JohnSmith-LT I committed the change from padding-bottom to padding-top. Looks ok to me but if you think there's a better solution then just revert my change.

**Comment by JohnSmith-LT** [comment]
@ronm123 the formatting looks much better.

Is this too wordy, does it mention bullets too often and/or do you think we should put something in about the fact that you can do numbered bullets and sub lists ?

![image](https://user-images.githubusercontent.com/3907362/69152254-00bdba80-0ad4-11ea-98e9-4e1621212a96.png)

**Comment by ronm123** [comment]
@JohnSmith-LT I think it's probably worth adding both tooltips and then splitting the main guidance for each e.g.

Page text:
If you want a mix of normal paragraph text (without bullets) and text that does have bullets (including numbered lists and sub lists) then apply the bullets and numbers to selected text using the wysiwyg editor buttons. Note: The option below to add bullets to all paragraphs will then be ignored.

Add bullets to text
If you wish to automatically add bullets to all paragraphs set this option to yes but do not add bullets via the wysiwyg editor. If you do also add bullets via the wysiwyg editor then this automatic option will be ignored.

**Comment by JohnSmith-LT** [comment]
The bullets page is now much more flexible than it was but is there anything else we need to add before we close this? The "timed text" reference could maybe be improved - i.e. we always had timed bullets but should we add an option to have different timings? Perhaps a comma separated list of different times to accomodate different length sections and/or to align transitions with narration?

Also #40 was about adding media support to the page but it's open to interpretation what that means - we obviously have media support in all those wysiwig boxes but it would be relatively easy to add other things into the sequence, but is that what was intended with #40 

Do you have any thoughts on this @ronm123 or @FayCross ?

**Comment by FayCross** [comment]
Is it worth branding this page under a different name as it's not really just a bullets page now? The name timed text would probably better describe it now. We could make the 'Add bullets to text' an optional property so it's not there by default on new pages but will still appear in editor & work as expected for any older ones. Personally I'm not a fan of really long tooltips to describe what things do and if we did this then the tooltips could be shortened. For example:

Page text: Each paragraph and bullet point (including numbered list and sub-list items) will appear separately at timed intervals on screen
Add bullets to text (now opt prop instead): Automatically convert all paragraphs in page text to bullet points

Other than that I'm happy for this to be closed - I don't think there's anything else that needs adding

**Comment by JohnSmith-LT** [comment]
Sure @FayCross always happy with the extra input. Work your magic and if you and @ronm123 are both happy then close it when are done.

**Comment by FayCross** [comment]
@ronm123 are you happy for this page to be renamed? Bullets isn't really the page's main purpose any more (& hasn't been for some time)

**Comment by JohnSmith-LT** [comment]
I'm still not 100% convinced that this shouldn't remain bullets but that we have a new Timed page which can have lots of things timed. In this page at present it is only really bullets and text outwith bulleted sections, although I suppose you could put images etc in the <p> blocks and they would animate too...

That said i'm definitely happy for it to be changed and then it could probably be improved upon.

**Comment by ronm123** [comment]
@FayCross @JohnSmith-LT a few comments...

**page name**
yes as long as the change of name is just a string and we keep the model file etc as it is so that no existing projects are broken then I think it does make more sense to call it timed text. Perhaps Timed text (Incl bullets) so that there is still reference to its previous name in guides and screenshots etc?

**tooltips**
Agree that they should be as short as is possible to explain the functionality. I would change Fays suggestion slightly:
Page text: Each paragraph and bullet point (including numbered list and sub-list items and images) will appear separately at timed intervals on screen. Apply bullets/numbers to selected text using the toolbar buttons.
Add bullets to text (now opt prop instead): Automatically add bullet points to all paragraphs. Note: this is ignored if any bullets/numbers are added via the toolbar buttons.

**additional timing features**
John mentioned possibly a comma separated list so there could possibly be different timings. I think an option for different timings would be useful/popular. However I'm not sure that would be very user friendly to have to count the paragraphs and then commas if for example you wanted the 4th para to be a different delay. Also then you have to be careful to update these values if you add or delete paragraphs. However what about optional nested pages? e.g. we have a number of pages where there is syncing to media but what if we had the same ability to add nested pages but instead of synching to media the sync points were just from the first/previous para loading. So what we have now where the first text works as it does with a default delay value but then optional nested pages where there is a wysiwyg text box and empty sync field where the default value is used unless a different value added but where that sync value is just delay from the previous content. An additional option might be to delay until return/space/click rather than automatic. Perhaps this is something that we should leave until a later date and discuss further or there's a better/easier way?

**Comment by ronm123** [comment]
I've created pages using the bullets page previously where I've included images that build too and the current functionality works quite well for this too if you are careful about where the image is added and what alignment is used. So this is already arguably 'Timed content' rather than just 'Timed text'.

There are a few ideas and possible functionality converging here though e.g. what about an optional delay property for most/all pages where an author could set a delay before each paragraph (or div or nested content) builds?

**Comment by ronm123** [comment]
Just thinking out loud here but worth adding while we are discussing this I think...

Historically we've had lots of different page types for different purposes. As we know over time and for various reasons many of them have become redundant or have far more flexibility than originally available e.g. the plain text page hasn't been just a plain text page for a long time! Likewise we've had separate navigators and then the multi-navigator which I'm a big fan of because it makes it so quick and easy to change from one type to another just by selecting the option. So I wonder about keeping this as the bullets page with the current improvements but then adding new options to the multi-navigator where one of those could be timed delay of each nested content?

**Comment by FayCross** [comment]
Timed content seems like a good name Ron. Do you think that if we are changing the name then it should wait until next release rather than be cherry picked into 3.8? It feels like something that should be in release notes rather than just added as if a bug fix.

**Comment by ronm123** [comment]
@FayCross well it's quickly evolved into a more flexible and powerful page and that was one of the reasons for the longer tooltip text given that if cherry picked for an incremental update it wouldn't be part of the 3.8 release notes. Perhaps we need to add additional sections (not pages) to release notes in future to cover incremental updates where new features require/justify additional guidance in the release notes.

John said: "I'm still not 100% convinced that this shouldn't remain bullets but that we have a new Timed page which can have lots of things timed." 

I'm thinking we could leave the Bullets page named as it is but with the current enhanced functionality and cherry pick that from 3.8.x without need for updated release notes but then discuss and work on a new timed content page, or possibly addition to the multi-navigator, or another existing page, or as I mentioned perhaps as an optional property to all relevant pages.

Ultimately I think what people want is more control over timed content for all relevant pages e.g. more animation/transition rather than just all content loading in one go unless using one of the synced media pages.

**Comment by JohnSmith-LT** [comment]
> I'm thinking we could leave the Bullets page named as it is but with the current enhanced functionality and cherry pick that from 3.8.x without need for updated release notes but then discuss and work on a new timed content page, or possibly addition to the multi-navigator, or another existing page, or as I mentioned perhaps as an optional property to all relevant pages.

I like the idea of making the multi-navigator the timed page...

**Comment by FayCross** [comment]
Forgot to tag issue in commit: https://github.com/thexerteproject/xerteonlinetoolkits/commit/3b2d2847a1b111a419e5ab32d508e7dad3f47052

I've changed a couple of the default settings to better suit the new functionality & renamed page for now to Bullets / Timed Content. We can always change this if & when a new fully functioning timed content page is made
